### DISCUSSION TITLE
Determining target framework in source generators

### DISCUSSION THREAD

**Discussion by normj** [dotnet/roslyn#72069] [Q&A]
I'm trying to understand from a source generator how to get the current target framework. I need to know the target framework to map a setting in an auxiliary file.  

I see in the [source-generators.cookbook.md](https://github.com/dotnet/roslyn/blob/a3c0cbd96fa27cded000dacb8114a7ab3d475420/docs/features/source-generators.cookbook.md?plain=1#L1123C1-L1123C156) there is a open issue showing how to do this but I can't find an example on how to do it.

```
**Feature detection**: Show how to create a generator that relies on specific target framework features, without depending on the TargetFramework property.
```

My best guess is we deliver a props like this to expose it but if that is direction I don't see how you would reference the props file in the `CSharpSourceGeneratorTest` for unit tests.
```
<Project>
  <ItemGroup>
    <CompilerVisibleProperty Include="TargetFramework" />
    <CompilerVisibleProperty Include="TargetFrameworks" />
  </ItemGroup>
</Project>
```

**Comment by normj** [comment]
We have figured out we can use `context.AnalyzerConfigOptions.GlobalOptions.TryGetValue("build_property.TargetFramework", out var targetFramework)` to get the target framework but it only seems to work when use it in a "real" scenario like from the NuGet package. It always returns back null when we try to use the `CSharpSourceGeneratorTest` from [Microsoft.CodeAnalysis.Testing](https://github.com/dotnet/roslyn-sdk/tree/main/src/Microsoft.CodeAnalysis.Testing#microsoftcodeanalysistesting)

**Comment by huoyaoyuan** [comment]
Target framework isn't a thing at the view of compiler. It can be compiled against a random set of reference libraries.

The common practice to detect runtime feature is detecting the **existence** of certain member. The constants defined in `System.Runtime.CompilerServices.RuntimeFeature` is for such usage.
